lib/src/st2110/pipeline/st40_pipeline_tx.h

```diff
@@ -0,0 +1,204 @@
+   * And only non-block method can be used within this callback as it run from lcore
+   * tasklet routine.
+   */
+  int (*notify_frame_available)(void* priv);
+  /**
+   * Optional. Callback when frame done.
+   * And only non-block method can be used within this callback as it run from lcore
+   * tasklet routine.
+   */
+  int (*notify_frame_done)(void* priv, struct st40_frame_info* frame);
+
+  /**
+   * Optional. tx destination mac address.
+   * Valid if ST40P_TX_FLAG_USER_P(R)_MAC is enabled
+   */
+  uint8_t tx_dst_mac[MTL_SESSION_PORT_MAX][MTL_MAC_ADDR_LEN];
+};
+
+struct st40p_tx_ctx {
+  struct mtl_main_impl* impl;
+  int idx;
+  int socket_id;
+  enum mt_handle_type type; /* for sanity check */
+
+  char ops_name[ST_MAX_NAME_LEN];
+  struct st40p_tx_ops ops;
+
+  st40_tx_handle transport;
+  uint16_t framebuff_cnt;
+  uint16_t framebuff_producer_idx;
+  uint16_t framebuff_consumer_idx;
+  struct st40p_tx_frame* framebuffs;
+  pthread_mutex_t lock;
+  bool ready;
+
+  int frames_per_sec;
+
+  /* for ST40P_TX_FLAG_BLOCK_GET */
+  bool block_get;
+  pthread_cond_t block_wake_cond;
+  pthread_mutex_t block_wake_mutex;
+  uint64_t block_timeout_ns;
+
+  /* get frame stat */
+  int stat_get_frame_try;
+  int stat_get_frame_succ;
+  int stat_put_frame;
+};
+
+/** Handle to tx st2110-40 pipeline session of lib */
+typedef struct st40p_tx_ctx* st40p_tx_handle;
+
+/** Bit define for flags of struct st40p_tx_ops. */
+enum st40p_tx_flag {
+  /**
+   * Flag bit in flags of struct st40_tx_ops.
+   * P TX destination mac assigned by user
+   */
+  ST40P_TX_FLAG_USER_P_MAC = (MTL_BIT32(0)),
+  /**
+   * Flag bit in flags of struct st40_tx_ops.
+   * R TX destination mac assigned by user
+   */
+  ST40P_TX_FLAG_USER_R_MAC = (MTL_BIT32(1)),
+  /**
+   * Flag bit in flags of struct st40_tx_ops.
+   * User control the frame pacing by pass a timestamp in st40_tx_frame_meta,
+   * lib will wait until timestamp is reached for each frame.
+   */
+  ST40P_TX_FLAG_USER_PACING = (MTL_BIT32(3)),
+  /**
+   * Flag bit in flags of struct st40_tx_ops.
+   * If enabled, lib will assign the rtp timestamp to the value in
+   * st40_tx_frame_meta(ST10_TIMESTAMP_FMT_MEDIA_CLK is used)
+   */
+  ST40P_TX_FLAG_USER_TIMESTAMP = (MTL_BIT32(4)),
+  /**
+   * Flag bit in flags of struct st40_tx_ops.
+   * If enable the rtcp.
+   */
+  ST40P_TX_FLAG_ENABLE_RTCP = (MTL_BIT32(5)),
+  /**
+   * Flag bit in flags of struct st40_tx_ops.
+   * If use dedicated queue for TX.
+   */
+  ST40P_TX_FLAG_DEDICATE_QUEUE = (MTL_BIT32(6)),
+  /** Force the numa of the created session, both CPU and memory */
+  ST40P_TX_FLAG_FORCE_NUMA = (MTL_BIT32(8)),
+  /** Enable the st40p_tx_get_frame block behavior to wait until a frame becomes
+   available or timeout(default: 1s, use st40p_tx_set_block_timeout to customize)*/
+  ST40P_TX_FLAG_BLOCK_GET = (MTL_BIT32(15)),
+};
+
+struct st40p_tx_frame {
+  enum st40p_tx_frame_status stat;
+  struct st40_frame_info frame;
```

Comment:
Mby call this frame_info 

reading the code the structures for frambuffers are called "frames"
which suggest st40p_tx_frame 
Then the frame_info structures are called ... frame 


This is very confusing to read i would love to change all instances to frame_info to make it apparent when you read the code

Reply:
renamed all `frame` to `frame_info`